### DISCUSSION TITLE
How to change the size of TopBottomPanel?

### DISCUSSION THREAD

**Discussion by singalen** [emilk/egui#949] [Q&A]
So far, I can only see `default_height()` method, but it doesn't change the current size.

AFAIU, I need to access `Memory`. I'm trying this:
```
            let id = egui::Id::new("game log");
            ui.memory().id_data.remove(&id);
```
but it doesn't help.

Maybe because I can only call it on `Ui` instance, and I only have access to it in a callback from`TopBottomPanel::show()`, and `TopBottomPanel::show()` at the end re-saves the data?

**Comment by parasyte** [comment]
edit: It looks like patching `TopBottomPanel` could help in the meantime. Since it remembers state between frames but doesn't provide access to it. It would use `PanelState` to update the rect, just like the `show` method does. You cannot access `Memory` from the outside because `PanelState` is private to the `egui` containers module.

Original answer is below, but I think I misinterpreted your question.

----

You cannot change the size of something after it has been drawn. This is a fundamental limit to immediate mode GUIs. You will have to figure out the size you want ahead of time, then create the `TopBottomPanel` with that height.

Note that #843 _might_ change that in the future. But the API and everything that can be supported is still TBD.

**Comment by singalen** [comment]
Thank you. Indeed, I want to change its size in a next frame, not after it was drawn in this frame.

I tried to completely wipe its Memory by the ID, but it didn't help. I would be fine if it worked.

**Reply by parasyte** [reply]
You won't be able to touch the memory without having access to the type that is stored. I recommend creating a PR that adds the feature.

**Reply by singalen** [reply]
Thanks, will do.

I will look into creating a trait, e.g. `Resizeable`. In general, I would add my voice to #539.

**Reply by singalen** [reply]
One attempt failed because it required double-locking the Memory: https://github.com/singalen/egui/commit/fdd617555a84b911bb97da58eca11761d54ab595

**Reply by singalen** [reply]
Here's a dirty hack, but I cannot in good conscience submit it as a PR: https://github.com/singalen/egui/commit/5a6eb28f834d55deb80f6680c61d76189361752c
and I don't see a cleaner way to do it 8(

**Reply by parasyte** [reply]
It's unclear to me how your trait is supposed to be used, but I was able to call resize from inside the `show` method. I adapted the `TopBottomPanel` doctest:

```rust
use egui::containers::panel::Resizable;

egui::TopBottomPanel::top("my_panel").show(ctx, |mut ui| {
    ui.label("Hello World!");
    egui::TopBottomPanel::resize(&mut ui, egui::Id::new("my_panel"), egui::Pos2::new(0.0, 100.0));
});
```

It's an unusual API, but I understand the limitations that lead to this design. Were you having problems calling it like this, or was the issue something else entirely?

**Reply by singalen** [reply]
No real reason for a trait here, I was just thinking how a more unified API would look (again, referring to #539).

Standard library, for example, has a trait for most operations that can be called on different "classes", like `IntoIterator`, `Extend`, `Sum` and a number of others. A separate trait a) unifies the APIs and b) sends a message about what operations can and what can not be performed on a certain type.

Attempt 1 didn't run for me, because `ui.memory()` tries to take a lock that is already taken by `show()`.
I don't see how it could have worked.

**Reply by parasyte** [reply]
> Attempt 1 didn't run for me, because `ui.memory()` tries to take a lock that is already taken by `show()`.
> I don't see how it could have worked.

It should be fine because the lock is dropped right away: https://github.com/emilk/egui/blob/214d2b5d4f10d58c65593fec93fce305e5d7c4fc/egui/src/containers/panel.rs#L27-L29 The `get_persisted` call will clone the value so the lock doesn't need to be held: https://github.com/emilk/egui/blob/214d2b5d4f10d58c65593fec93fce305e5d7c4fc/egui/src/util/id_type_map.rs#L360-L366

If it didn't release the lock, then it wouldn't be able to acquire it again later: https://github.com/emilk/egui/blob/214d2b5d4f10d58c65593fec93fce305e5d7c4fc/egui/src/containers/panel.rs#L490 (this lock is dropped at the end of the block)...

**Reply by singalen** [reply]
Oh, I see. I did it on 0.15 release, because I use it through `egui-macroquad`.

**Reply by parasyte** [reply]
It's the same on 0.15, just with some helper methods.

**Reply by singalen** [reply]
Then I must be missing something, because for me it crashes with the following stack:
```
   4: epaint::mutex::mutex_impl::Mutex<T>::lock
             at ./depends/egui/epaint/src/mutex.rs:166:13
   5: egui::context::Context::memory
             at ./depends/egui/egui/src/context.rs:375:9
   6: egui::ui::Ui::memory
             at ./depends/egui/egui/src/ui.rs:319:9
   7: <egui::containers::panel::TopBottomPanel as egui::containers::panel::Resizable>::resize
             at ./depends/egui/egui/src/containers/panel.rs:446:13
   8: rifrl::systems::platform::macroquad::ui_render::ui_render::{{closure}}::{{closure}}
             at ./src/systems/platform/macroquad/ui_render.rs:286:17
   9: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
  10: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/alloc/src/boxed.rs:1691:9
  11: egui::containers::panel::TopBottomPanel::show_inside_dyn::{{closure}}
             at ./depends/egui/egui/src/containers/panel.rs:537:13
```
(line numbers are somewhat off, because I have a few lines added in `panel.rs`)

**Reply by parasyte** [reply]
Maybe you could show the full patch you are using? When I tested the sample code I pasted earlier, I literally made the change to the existing doctest against your branch https://github.com/singalen/egui/commit/fdd617555a84b911bb97da58eca11761d54ab595 and the test passed.

Also, it might help knowing exactly what the error is; did the mutex get poisoned because your code panicked while the lock was held? Or did you introduce reentrancy that causes the `lock` call to panic?

**Reply by singalen** [reply]
```
at 'already mutably borrowed', /Users/vic/.cargo/registry/src/github.com-1ecc6299db9ec823/atomic_refcell-0.1.8/src/lib.rs:151:23
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::panicking::panic_display
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:64:5
   3: atomic_refcell::AtomicRefCell<T>::borrow_mut
             at /Users/vic/.cargo/registry/src/github.com-1ecc6299db9ec823/atomic_refcell-0.1.8/src/lib.rs:151:23
   4: epaint::mutex::mutex_impl::Mutex<T>::lock
             at ./depends/egui/epaint/src/mutex.rs:166:13
```

Diffs: https://github.com/singalen/egui/tree/0-15-patched

Usage:

```
        egui::TopBottomPanel::new(egui::panel::TopBottomSide::Top, "game log")
            .default_height(ui_log_height)
            .frame(egui::Frame::none().fill(Color32::BLACK).multiply_with_opacity(0.7))
            // .resize(Some(ui_log_height))
            .show(egui_ctx, |ui| {
                <TopBottomPanel as Resizable>::resize(ui, my_id, egui::Pos2::new(0., ui_log_height));

// ...

            });
```

**Reply by parasyte** [reply]
Ok, you're right! The extra context helped me reproduce the issue.

The relevant line of the backtrace is here: https://github.com/singalen/egui/blob/0-15-patched/egui/src/containers/panel.rs#L446 There is only one way this could possibly attempt to acquire a lock that is already acquired. If the lock on line 442 is still held!

The fix is simple enough, you just need to make sure that you drop the lock before calling into the condition body:

```rust
impl Resizable for TopBottomPanel {
    fn can_resize_y(&self) -> bool {
        true
    }

    fn resize(ui: &mut Ui, my_id: Id, new_dimensions: Pos2) {
        let state = {
            // This lock will be dropped at the end of block scope
            let mut guard = ui.memory();

            guard.id_data.get::<PanelState>(&my_id).copied()
        };
        if let Some(state) = state {
            let mut new_rect = state.rect;
            new_rect.set_height(new_dimensions.y);
            //PanelState { rect: new_rect }.store(ui.ctx(), my_id);
            ui.memory()
                .id_data
                .insert(my_id, PanelState { rect: new_rect });
        }
    }
}
```

There are several ways to do this. What I've done here is bind the lock guard to a variable that I know will be dropped at the end of the enclosing block scope. You can just as easily use an explicit `drop(guard)`:

```rust
impl Resizable for TopBottomPanel {
    fn can_resize_y(&self) -> bool {
        true
    }

    fn resize(ui: &mut Ui, my_id: Id, new_dimensions: Pos2) {
        // This lock must be dropped ASAP
        let mut guard = ui.memory();
        let state = guard.id_data.get::<PanelState>(&my_id).copied();
        drop(guard);

        if let Some(state) = state {
            let mut new_rect = state.rect;
            new_rect.set_height(new_dimensions.y);
            //PanelState { rect: new_rect }.store(ui.ctx(), my_id);
            ui.memory()
                .id_data
                .insert(my_id, PanelState { rect: new_rect });
        }
    }
}
```

Either of these will work.

It also occurs to me that the reason I was unable to reproduce this issue is because the doctest only runs the `show` method one time. That isn't enough to populate the memory and then trigger the bug. Simply running the same code twice in sequence does it. So I was able to both confirm your observations and test that dropping the lock guard in your implementation fixes it.

**Reply by singalen** [reply]
Thanks a ton!

Well, it still didn't work for me. I guess because this `resize(ui, ...)` would be called from inside `show()`, and `show()` makes a copy of the `PanelState` at the beginning and saves it in the end. This overwrites whatever changes the user could make from `add_contents()`.

I am, so far, going with my local ugly hack: https://github.com/singalen/egui/commit/5a6eb28f834d55deb80f6680c61d76189361752c

I don't see a good way to fit it into anything that resembles a designed API on `TopBottomPanel`.

**Comment by singalen** [comment]
In the end, I did everything with `Window`s.
